File: problem/binarytrees/solution/swift/4.swift
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// based on programs by Marcel Ibes and Ralph Ganszky
// contributed by Isaac Gouy


import Dispatch
import Foundation

class TreeNode {
    var left, right: TreeNode?

    init(left: TreeNode?, right: TreeNode?) {
        self.left = left
        self.right = right
    }

    func check() -> Int {
        if left != nil {
            return left!.check() + right!.check() + 1
        } else {
            return 1
        }
    }
}

func bottomUpTree(_ depth: Int) -> TreeNode? {
    if depth > 0 {
        let node = TreeNode(left: bottomUpTree(depth-1),
                            right: bottomUpTree(depth-1))
        return node
    } else {
        let node = TreeNode(left: nil, right: nil)
        return node
    }
}

func inner(depth: Int, iterations: Int) -> String {
    var chk = 0
    for _ in 0..<iterations {
        let a = bottomUpTree(depth)
        chk += a!.check()
    }
    return "\(iterations)\t trees of depth \(depth)\t check: \(chk)"
}

let n: Int

if CommandLine.argc > 1 {
    n = Int(CommandLine.arguments[1]) ?? 10
} else {
    n = 10
}

let minDepth = 4
let maxDepth = (n > minDepth + 2) ? n : minDepth + 2
var messages: [Int:String] = [:]
let depth = maxDepth+1

let group = DispatchGroup()

let workerQueue = DispatchQueue.init(label: "workerQueue", qos: .userInitiated,
attributes: .concurrent)
let messageQueue = DispatchQueue.init(label: "messageQueue", qos: .userInitiated
)

group.enter()
workerQueue.async {
    let tree = bottomUpTree(depth)

    messageQueue.async {
        messages[0] = "stretch tree of depth \(depth)\t check: \(tree!.check())"
        group.leave()
    }
}

group.enter()
workerQueue.async {
    let longLivedTree = bottomUpTree(maxDepth)

    messageQueue.async {
        messages[Int.max] = "long lived tree of depth \(maxDepth)\t check: \(lon
gLivedTree!.check())"
        group.leave()
    }
}

group.enter()
workerQueue.async {
    for halfDepth in (minDepth / 2)..<(maxDepth/2+1) {
        let depth = halfDepth * 2
        let iterations = 1 << (maxDepth - depth + minDepth)

        group.enter()
        workerQueue.async {
            let msg = inner(depth: depth, iterations: iterations)
            messageQueue.async {
                messages[depth] = msg
                group.leave()
            }
        }
    }

    messageQueue.async {
        group.leave()
    }
}

group.wait()

for msg in messages.sorted(by: { $0.0 < $1.0 }) {
    print(msg.value)
}

File: problem/binarytrees/solution/rust/4.rs
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// contributed by the Rust Project Developers
// contributed by TeXitoi
// contributed by Cristi Cobzarenco
// contributed by Matt Brubeck
// modified by Tom Kaitchuck
// modified by Volodymyr M. Lisivka

extern crate rayon;
extern crate toolshed;

use toolshed::Arena;
use rayon::prelude::*;

#[derive(Debug, PartialEq, Clone, Copy)]
struct Tree<'a> {
    left: Option<&'a Tree<'a>>,
    right: Option<&'a Tree<'a>>,
}

fn item_check(tree: &Tree) -> i32 {
    if let (Some(left), Some(right)) = (tree.left, tree.right) {
        1 + item_check(right) + item_check(left)
    } else {
        1
    }
}

fn bottom_up_tree<'r>(arena: &'r Arena, depth: i32)
                  -> &'r Tree<'r> {
    let tree = arena.alloc(Tree { left: None, right: None });
    if depth > 0 {
        tree.right = Some(bottom_up_tree(arena, depth - 1));
        tree.left = Some(bottom_up_tree(arena, depth - 1));
    }
    tree
}

fn inner(depth: i32, iterations: i32) -> String {
    let chk: i32 = (0 .. iterations).into_par_iter().map(|_| {
        let arena = Arena::new();
        let a = bottom_up_tree(&arena, depth);
        item_check(a)
    }).sum();
    format!("{}\t trees of depth {}\t check: {}", iterations, depth, chk)
}

fn main() {
    let n = std::env::args().nth(1)
        .and_then(|n| n.parse().ok())
        .unwrap_or(10);
    let min_depth = 4;
    let max_depth = if min_depth + 2 > n { min_depth + 2 } else { n };

    {
        let arena = Arena::new();
        let depth = max_depth + 1;
        let tree = bottom_up_tree(&arena, depth);
        println!("stretch tree of depth {}\t check: {}", depth, item_check(tree)
);
    }

    let long_lived_arena = Arena::new();
    let long_lived_tree = bottom_up_tree(&long_lived_arena, max_depth);

    let messages = (min_depth/2..max_depth/2 + 1).into_par_iter().map(|half_dept
h| {
            let depth = half_depth * 2;
            let iterations = 1 << ((max_depth - depth + min_depth) as u32);
            inner(depth, iterations)
        }).collect::<Vec<_>>();

    for message in messages {
        println!("{}", message);
    }

    println!("long lived tree of depth {}\t check: {}", max_depth, item_check(lo
ng_lived_tree));
}

